content/blog/2019-12-20-Internal-Function-Memoization/index.md

```diff
@@ -0,0 +1,175 @@
+
+Consider the layout of the iterative version of the same program below.
+
+<img src="memo.png" style="max-width: 100%" >
+
+Note: We keep the function header the same in order to more easily swap out invocations to fibonacci outside the function with our generated function.
+
+### New Code Generation
+To minimize changes to the overall code structure, we delete all blocks and instructions inside the original recursive function and insert our new instructions.
+
+First we add all the base case conditionals. Next, we declare and initialize all the values we need for iteration and the incrementing iterator. After that, we create the while loop and clone in the instructions dependent on the original recursive calls (e.g., the add instruction in fib(n-1) + fib(n-2)). Lastly, we add the return statementment to which the while loop exits. 
+
+Our code for the LLVM pass is available [here](https://github.com/liuhenry4428/llvm-pass-skeleton/tree/noauto).
+
+<!-- We start of by 
+#### Fibonacci
+Consider the fibonacci algorithm. To compute number $n$, you only need the numbers (n-1) and (n-2). Any additional previous calculations can be thrown away. Is it possible to infer this information when given a program, such that we can alter a program to decrease its space consumption? If so, what are the cases to make this more generalizable, beyond optimizing a fiboncci program?
+
+Consider the call tree for recursive fibonacci.
+![](https://i.imgur.com/xaaXLDV.png)
+
+
+![](https://i.imgur.com/uCtIcSt.png)
+
+How extensible are these ideas? Consider the arbitrary one-dimensional programming problem below. Without specific knowledge of which recursive calls are required and their offsets from $n$, we must conservatively estimate a next call could use any previously computed elements.
+
+## Evaluation
+### Correctness
+In order to evaluate correctness, we look at the programs generated for the set of valid benchmarks. During construction, we separate the domain by case into separate functions, each with a different result. These _separate functions_ are the base cases, each with their own code that exeutes and returns within the case. The \{domain $\setminus$ the domain of the base cases\} is input into the loop in the body of our function. Determining program equivalence can be broken down into equivalence of these separate functions.
+
+However, function equivalence is an undecidable problem, as that would solve the [halting problem](https://en.wikipedia.org/wiki/Halting_problem). Therefore, to approximately validate our generated programs are correct, we test the bases cases, and a few arguments for the iterative case. 
+We randomly generating a few integers between 2 and 35 as 1 is a base case, and 35 is the largest fibonacci number that still fits in a 32-bit integer.
```

Comment:
generating -> generate

Reply:
If the inputs from 1 through 35 are the only valid ones, then exhaustive testing should be tractable. :smiley: